### DISCUSSION TITLE
Calling PropertyChanged events from worker threads

### DISCUSSION THREAD

**Discussion by ArchibaldMacdonald-cyacomb** [AvaloniaUI/Avalonia#15286] [Q&A]
I'm having trouble finding official information on calling `PropertyChanged` events from off the main UI-thread. I've tracked down 2 github issues that suggest one shouldn't, but I'm not sure how in-date these are with the current state of Avalonia framework:

https://github.com/AvaloniaUI/Avalonia/issues/7644
https://github.com/AvaloniaUI/Avalonia/issues/5550

So a couple of questions:

1. Definitively, is it still the case that can you not call `PropertyChanged` off-thread?
1. Why does it _currently_ work in cases I've witnessed, and is this undefined behaviour?
1. Can we have official documentation on this for the future?
1. Can we have warnings/exceptions in the future such that it doesn't silently "just work" or is this by design?

**Comment by timunie** [comment]
3 --> https://docs.avaloniaui.net/docs/guides/development-guides/accessing-the-ui-thread

**Reply by ArchibaldMacdonald-cyacomb** [reply]
Hi Timunie,

The docs you linked talk about how to access the UI thread, but don't give much info of when and why.

**Reply by timunie** [reply]
everything that updates the UI, so all events the UI needs to handle. Feel free to add a PR to the docs to clarify it further.

**Reply by ArchibaldMacdonald-cyacomb** [reply]
Appreciate it Timunie, I needed it written down explicitly. (And answers questions 1 and 3.)

**Comment by robloo** [comment]
Long-term there might be things that Avalonia should do to be more similar to WPF here: https://stackoverflow.com/questions/1321423/does-wpf-databinding-marshall-changes-to-the-ui-thread

**Comment by chenjing1294** [comment]
@ArchibaldMacdonald-cyacomb Hi, Sometimes I call it successfully in other threads, which may be related to this, https://github.com/AvaloniaUI/Avalonia/commit/8280b83e95d1f6803194a27bd3c601cf120b21de

**Reply by thevortexcloud** [reply]
You may actually be on the UI thread. Threading in .NET is weird. EG if you start an async task, you are not guaranteed to return back to the same thread which started and awaited the task. You sort of just end up on any thread that's available.

**Reply by ArchibaldMacdonald-cyacomb** [reply]
@thevortexcloud this is true if you use `.ConfigureAwait(false)` in some circumstances. But for the default `.ConfigureAwait(true)` case, you are at least guaranteed to return to the same synchronisation context - and the UI thread has its [own context](https://reference.avaloniaui.net/api/Avalonia.Threading/AvaloniaSynchronizationContext/).
